services/sysinfo/server/sysinfo.go

```diff
@@ -57,6 +62,58 @@ func (s *server) Uptime(ctx context.Context, in *emptypb.Empty) (*pb.UptimeReply
 	return reply, nil
 }
 
+func (s *server) Dmesg(request *pb.DmesgRequest, stream pb.SysInfo_DmesgServer) error {
+	ctx := stream.Context()
+	// logger := logr.FromContextOrDiscard(ctx)
+	recorder := metrics.RecorderFromContextOrNoop(ctx)
+	if request.Grep == "" && (request.IgnoreCase || request.InvertMatch) {
+		return status.Error(codes.InvalidArgument, "must provide grep argument before setting ignore_case or invert_match")
+	}
+
+	records, err := getKernelMessages()
+	if err != nil {
+		recorder.CounterOrLog(ctx, sysinfoDmesgFailureCounter, 1, attribute.String("reason", "get kernel message error"))
+		return status.Errorf(codes.InvalidArgument, "can't get kernel message %v", err)
+	}
+	// grep the messages we want
+	if request.Grep != "" {
+		var newRecords []*pb.DmsgRecord
+		regex := request.Grep
+		if request.IgnoreCase {
+			regex = `(?i)` + regex
+		}
+
+		isInvert := request.InvertMatch
+		pattern := regexp.MustCompile(regex)
+		for _, record := range records {
+			// Two cases are what we want
+			// case 1: isInvert = false and match = true (pattern matched)
+			// case 2: isInvert = true and match = false (pattern doesn't match)
+			if match := pattern.MatchString(record.Message); match != isInvert {
+				newRecords = append(newRecords, record)
+			}
+		}
+		records = newRecords
+	}
+
+	// tail the last n lines
+	// there is no way to tail number of messages more than initial records
+	if request.TailLines > int32(len(records)) {
+		request.TailLines = int32(len(records))
+	}
+	// negative number means disables the tail feature
+	if request.TailLines >= 0 {
```

Comment:
I'd minority prefer making this `request.TailLines > 0`. 0 is the default value (https://protobuf.dev/programming-guides/proto3/#default), so returning no data when tail_lines is unset is a little unexpected.

Reply:
That makes sense to me.